### ISSUE TITLE
Documentation about PUSH-notifications

### ISSUE THREAD

**Issue by SomeGeek** [BelledonneCommunications/linphone-android#805]
Is there any documentation about push notifications available for Linphone?

What should be the right content of such a message to 'wake up' Linphone? And would it be a good idea to have a option inside the UI to define your FCM tokens? 

For clarity: I'm talking about push notifications sent from my own PBX, _not_ sip.linphone.org!

**Comment by SomeGeek** [comment]
Yep, it's how to _include your own FCM keys_. It doesn't mention anything about the _server side_ of things. What kind of notification do I need to send?

[This piece of code](https://github.com/BelledonneCommunications/flexisip/blob/23af6b202bb9b74c865d2dce86eb65eb69fc4a96/src/module-pushnotification.cc#L209) shines some light on the subject.

```
- $type      : apple, google, wp, firebase\n"
- $token     : device token\n"
- $api-key   : the api key to use (google and firebase only)\n"
- $app-id    : application ID\n"
- $from-name : the display name in the from header\n"
- $from-uri  : the sip uri of the from header\n"
- $from-tag  : the tag of the from header \n"
- $to-uri    : the sip uri of the to header\n"
- $call-id   : the call-id of the INVITE or MESSAGE request\n"
- $event     : call, message\n"
- $sound     : the sound file to play with the notification\n"
- $msgid     : the message id to put in the notification\n"
- $uid       : \n"
```

But how does this need to be placed inside such a request?

**Comment by Viish** [comment]
That a look at flexisip pusher tool : https://github.com/BelledonneCommunications/flexisip/blob/23af6b202bb9b74c865d2dce86eb65eb69fc4a96/src/tools/pusher.cc
Also you can look at the firebase documentation because flexisip code won't help you much to do the work in Asterisk...

**Comment by SomeGeek** [comment]
Let's distinguish three things here.

- The SIP server (the PBX in this case)
- The SIP client (Linphone running on Android)
- The push-notification delivery service (FCM)

The only reason I looked into Flexisip code is to find out what kind of information I need to have _inside_ my push message. At _some_ point in the dialplan, I need to sent out a push message. How to sent a push message is out of scope for this question.

Linphone recognizes some specific push message as a wakeup call. When I get to send the right message (and right referring to right content _of_ that message), it will do so. After that, it's a SIP configuration matter.

This pusher tool seems to differ from the Flexisip one, as I don't see any SIP addresses being sent out. Is that correct?

**Comment by Viish** [comment]
For Android we don't care about the content of the push, it only has to be a high priority data message.

**Comment by sercan447** [comment]
@SomeGeek 
hello, what is the solution for this problem?
how to wake app with firebase notification?
do you have a sample json format?

**Comment by Viish** [comment]
@sercan447 take a look at our wiki https://wiki.linphone.org/xwiki/wiki/public/view/Lib/Features/Push%20notifications/

**Comment by sercan447** [comment]
@Viish 

I can wake up my linphone application with this data format that I have thrown with firebase.

{
    "to": "x7NrNknM8thlNA3PDZrjf0ETggm2Tvxq_hvI20QTk0uRwuoR5MgXUDYNU",
    "time_to_live": 0,
    "priority": "high",
    "data": {
        "uuid": "<uid>",
        "form-uri": "<from-uri>",
        "display-name": "<from-name>",
        "call-id": "<call-id>",
        "sip-from": "<from>",
        "loc-key": "",
        "loc-args": "<from>",
        "send-time": "<date>"
    }
}


I don't know what to fill in the data like "form-uri","sip-from" here as an example. can you help me?

**Comment by Viish** [comment]
@sercan447 For Android these information aren't used, only call-id is used.
If it matches the call-id of the message or call that triggered the push sending and that will be received when the Core will be waken up it will terminate the background task faster than if this information isn't set or doesn't match the reality.

**Comment by sercan447** [comment]
here just sending call-Id didn't solve my problem. I'm passing the rest of the parameters empty.

{
    "to": "8thlNA3PDZrjf0ETggm2Tvxq_hvI20QTk0uRwuoR5MgXUDYNUyJ-KMprbluyoy5FX63Pl-uP9HyQFHFidLm3vkZSh4IQzY",
    "time_to_live": 0,
    "priority": "high",
    "data": {
        "uuid": "<uid>",
        "form-uri": "<from-uri>",
        "display-name": "<from-name>",
        **"call-id": "abd77cd7-48e0-4baf-bcee-81ffcb3a78c2",**
        "sip-from": "<from>",
        "loc-key": "",
        "loc-args": "<from>",
        "send-time": "<date>"
    }
}

**Comment by sercan447** [comment]
I initiate a call while the application is closed, that is, "quit". When the call is coming, I take the "call-id" and send it to firebase with postman.
My linphone app foreground service starts running. but it cannot find the incoming call and show it to me.

**Comment by Viish** [comment]
Please attach logs showing what happens when a push is received.

**Comment by sercan447** [comment]
Can we talk privately with you?
sercan.goger@gmail.com
can you reach me here?

**Comment by Viish** [comment]
@sercan447 you can send me your logs at sylvain.berfini@belledonne-communications.com

**Comment by Viish** [comment]
Your log stops suddenly when the Core starts during the registration process, and given the startup logs showing previous exit reasons it seems it crashes.
Can you catch the crash stacktrace using adb logcat tool?
Also why do you make LIME requests to our server? If that's a mistake, simply any mention of `lime.linphone.org` from your app.

**Comment by sercan447** [comment]
"Also why do you make LIME requests to our server?"  we don't do anything like that.
listen;


-We have two phones.
-linphone app is installed on both phones.
-Both phones are registered to the same account.
-Now we open one of the applications on the phone and close the other completely.
-then we look for the account where the phones are registered from a third account.
-When a call is made, the first phone receives the call directly because the linphone application is open. but the linphone app on the other phone is completely closed. and we take the "call-id" information from the first phone's logs via android studio and send it to the second phone as a notification.
-We want to continue the call when we click on the notification on the second phone.
-We are trying to do this with FREEBPX ip server.
-but this scenario I described above can only be done by flexisip. freepbx or similar servers with other sip servers cannot do this.

**Comment by Viish** [comment]
@sercan447 if your proxy server (FreePBX in your case) can't send push notifications itself, you can always install a flexisip as "push gateway". You can find a documentation about it here: https://wiki.linphone.org/xwiki/wiki/public/view/Flexisip/HOWTOs/Push%20Gateway/

**Comment by pergolafabio** [comment]
Guys, was reading this thread.... Am i correct saying that the pusher tool is a seperate tool? No flexisip is needed?

If yes, that means i can execute the tool to wake up linhome/linphone with a push message , with the original json supplied by the app? If yes, how can i execute this command? Any example is very welcome

thnx!!

**Comment by Viish** [comment]
@pergolafabio No, the pusher relies on flexisip so you can't have one without the other.

And for push to work, you need push certificates to match the ones your app has been build for.

**Comment by iamhalje** [comment]
@Viish Hello, after reading all the documentation/articles/maillists I found, it remains unclear to me whether it is possible to download the application from the "Play Store" and which API to send a request from my SIP proxy(without or with flexisip if this required), or for this I need to create a project in Firebase, rename the project, replace app/google-services .json, and only after that will I be able to send Push notifications to my client who use Linphone mobile app?

**Comment by Viish** [comment]
Hi @iamhalje,

You can't send push notifications from your server to our app on the Play Store because our server push certificate is private and we currently don't offer a public API to trigger push notifications from third party components.

You need to rebuild our app with your own google-services.json file that contains your own push certificate matching the one configured in your SIP server.

Cheers,
